File: backend/app/services/email_service.py
"""Email service for sending password reset codes and other notifications."""

from __future__ import annotations

import os  # APP_ENV + SMTP creds are read straight from env
import random  # generate_reset_code
import smtplib  # production SMTP transport
from email.mime.text import MIMEText  # HTML email body
from email.mime.multipart import MIMEMultipart  # multipart/alternative message
from typing import Optional

from app.core.config import settings


def generate_reset_code() -> str:
    """Generate a 6-digit numeric code."""
    return f"{random.randint(100000, 999999)}"


async def send_password_reset_email(email: str, code: str) -> bool:
    """
    Send a password reset code via email.
    
    In development, prints to console instead of sending.
    In production, uses SMTP settings from environment.
    """
    subject = "COURSER Password Reset Code"
    body = f"""
    <html>
      <body>
        <h2>Password Reset Request</h2>
        <p>You requested a password reset for your COURSER account.</p>
        <p>Your verification code is: <strong style="font-size: 24px; letter-spacing: 4px;">{code}</strong></p>
        <p>This code will expire in 15 minutes.</p>
        <p>If you didn't request this, please ignore this email.</p>
        <hr>
        <p style="color: #666; font-size: 12px;">COURSER Team</p>
      </body>
    </html>
    """

    # Development mode - print to console
    if os.getenv("APP_ENV", "production") == "development":
        print(f"\n{'='*50}")
        print(f"PASSWORD RESET EMAIL (dev mode)")
        print(f"To: {email}")
        print(f"Code: {code}")
        print(f"{'='*50}\n")
        return True

    # Production mode - send via SMTP
    smtp_host = os.getenv("SMTP_HOST")
    smtp_port = int(os.getenv("SMTP_PORT", "587"))
    smtp_user = os.getenv("SMTP_USER")
    smtp_password = os.getenv("SMTP_PASSWORD")
    from_email = os.getenv("FROM_EMAIL", "noreply@courser.app")

    # Missing any SMTP credential ⇒ cannot send; fail soft so the reset
    # flow still returns success to the client (no user enumeration).
    if not all([smtp_host, smtp_user, smtp_password]):
        print("WARNING: SMTP settings not configured, email not sent")
        return False

    try:
        # Build a multipart HTML message and send over STARTTLS.
        msg = MIMEMultipart("alternative")
        msg["Subject"] = subject
        msg["From"] = from_email
        msg["To"] = email
        msg.attach(MIMEText(body, "html"))

        with smtplib.SMTP(smtp_host, smtp_port) as server:
            server.starttls()  # upgrade to TLS before authenticating
            server.login(smtp_user, smtp_password)
            server.send_message(msg)
        return True
    except Exception as e:
        print(f"Failed to send email: {e}")
        return False


async def send_verification_email(email: str, code: str) -> bool:
    """Send an email verification code (same dev/prod split as above)."""
    subject = "COURSER Email Verification"
    body = f"""
    <html>
      <body>
        <h2>Verify Your Email</h2>
        <p>Welcome to COURSER! Please verify your email address.</p>
        <p>Your verification code is: <strong style="font-size: 24px; letter-spacing: 4px;">{code}</strong></p>
        <p>This code will expire in 15 minutes.</p>
        <hr>
        <p style="color: #666; font-size: 12px;">COURSER Team</p>
      </body>
    </html>
    """

    if os.getenv("APP_ENV", "production") == "development":
        print(f"\n{'='*50}")
        print(f"EMAIL VERIFICATION (dev mode)")
        print(f"To: {email}")
        print(f"Code: {code}")
        print(f"{'='*50}\n")
        return True

    smtp_host = os.getenv("SMTP_HOST")
    smtp_port = int(os.getenv("SMTP_PORT", "587"))
    smtp_user = os.getenv("SMTP_USER")
    smtp_password = os.getenv("SMTP_PASSWORD")
    from_email = os.getenv("FROM_EMAIL", "noreply@courser.app")

    if not all([smtp_host, smtp_user, smtp_password]):
        print("WARNING: SMTP settings not configured, email not sent")
        return False

    try:
        msg = MIMEMultipart("alternative")
        msg["Subject"] = subject
        msg["From"] = from_email
        msg["To"] = email
        msg.attach(MIMEText(body, "html"))

        with smtplib.SMTP(smtp_host, smtp_port) as server:
            server.starttls()
            server.login(smtp_user, smtp_password)
            server.send_message(msg)
        return True
    except Exception as e:
        print(f"Failed to send email: {e}")
        return False

File: backend-node/src/services/emailService.ts
/**
 * Email service for password reset codes — mirrors backend/app/services/email_service.py.
 *
 * In development the code is printed to the console; in production the email
 * is sent over SMTP via nodemailer. Missing SMTP credentials fail soft (the
 * reset flow still returns success so email addresses can't be enumerated).
 */
import crypto from "node:crypto";
import nodemailer from "nodemailer";

import { isDev } from "../config.js";

export function generateResetCode(): string {
  // crypto.randomInt is cryptographically strong (Math.random is predictable
  // and unsuitable for security codes).
  return String(crypto.randomInt(100000, 1000000));
}

function devLog(title: string, email: string, code: string): void {
  console.log("\n" + "=".repeat(50));
  console.log(`${title} (dev mode)`);
  console.log(`To: ${email}`);
  console.log(`Code: ${code}`);
  console.log("=".repeat(50) + "\n");
}

function htmlBody(subject: string, code: string, intro: string, expiresIn = "15 minutes"): string {
  return `
    <html>
      <body>
        <h2>${subject}</h2>
        <p>${intro}</p>
        <p>Your verification code is: <strong style="font-size: 24px; letter-spacing: 4px;">${code}</strong></p>
        <p>This code will expire in ${expiresIn}.</p>
        <p>If you didn't request this, please ignore this email.</p>
        <hr>
        <p style="color: #666; font-size: 12px;">COURSER Team</p>
      </body>
    </html>
  `;
}

async function send(
  email: string,
  code: string,
  subject: string,
  intro: string,
  devTitle: string,
  expiresIn = "15 minutes"
): Promise<boolean> {
  if (isDev()) {
    devLog(devTitle, email, code);
    return true;
  }

  const smtpHost = process.env.SMTP_HOST;
  const smtpPort = parseInt(process.env.SMTP_PORT || "587", 10);
  const smtpUser = process.env.SMTP_USER;
  const smtpPassword = process.env.SMTP_PASSWORD;
  const fromEmail = process.env.FROM_EMAIL || "noreply@courser.app";

  if (!smtpHost || !smtpUser || !smtpPassword) {
    console.warn("WARNING: SMTP settings not configured, email not sent");
    return false;
  }

  try {
    const transporter = nodemailer.createTransport({
      host: smtpHost,
      port: smtpPort,
      secure: smtpPort === 465,
      auth: { user: smtpUser, pass: smtpPassword },
    });
    await transporter.sendMail({
      from: fromEmail,
      to: email,
      subject,
      html: htmlBody(subject, code, intro, expiresIn),
    });
    return true;
  } catch (err) {
    console.error("Failed to send email:", err);
    return false;
  }
}

export async function sendPasswordResetEmail(email: string, code: string): Promise<boolean> {
  return send(
    email,
    code,
    "COURSER Password Reset Code",
    "You requested a password reset for your COURSER account.",
    "PASSWORD RESET EMAIL"
  );
}

export async function sendVerificationEmail(email: string, code: string): Promise<boolean> {
  return send(
    email,
    code,
    "COURSER Email Verification Code",
    "Confirm your email address to activate your COURSER account.",
    "EMAIL VERIFICATION CODE",
    "60 minutes"
  );
}